sys/usb/usbus/cdc/acm/cdc_acm.c

```diff
@@ -0,0 +1,306 @@
+                     sizeof(usb_desc_union_t) +
+                     sizeof(usb_desc_call_mngt_t),
+    },
+    .len_type = USBUS_HDR_LEN_FIXED,
+};
+
+static size_t _gen_mngt_descriptor(usbus_t *usbus)
+{
+    usb_desc_call_mngt_t mngt;
+    /* functional call management descriptor */
+    mngt.length = sizeof(usb_desc_call_mngt_t);
+    mngt.type = USB_TYPE_DESCRIPTOR_CDC;
+    mngt.subtype = 0x1;
+    mngt.capabalities = 0;
+    mngt.data_if = 1;
+    usbus_control_slicer_put_bytes(usbus, (uint8_t*)&mngt, sizeof(mngt));
+    return sizeof(usb_desc_call_mngt_t);
+}
+
+static size_t _gen_union_descriptor(usbus_t *usbus,
+                                    usbus_cdcacm_device_t *cdcacm)
+{
+    usb_desc_union_t uni;
+    /* functional union descriptor */
+    uni.length = sizeof(usb_desc_union_t);
+    uni.type = USB_TYPE_DESCRIPTOR_CDC;
+    uni.subtype = 0x6;
+    uni.master_if = cdcacm->iface_ctrl.idx;
+    uni.slave_if = cdcacm->iface_data.idx;
+    usbus_control_slicer_put_bytes(usbus, (uint8_t*)&uni, sizeof(uni));
+    return sizeof(usb_desc_union_t);
+}
+
+static size_t _gen_acm_descriptor(usbus_t *usbus)
+{
+    usb_desc_acm_t acm;
+    /* functional cdc acm descriptor */
+    acm.length = sizeof(usb_desc_acm_t);
+    acm.type = USB_TYPE_DESCRIPTOR_CDC;
+    acm.subtype = 0x02;
+    acm.capabalities = 0x00;
+    usbus_control_slicer_put_bytes(usbus, (uint8_t*)&acm, sizeof(acm));
+    return sizeof(usb_desc_acm_t);
+}
+
+static size_t _gen_cdc_descriptor(usbus_t *usbus)
+{
+    usb_desc_cdc_t cdc;
+    /* functional cdc descriptor */
+    cdc.length = sizeof(usb_desc_cdc_t);
+    cdc.bcd_cdc = 0x0120;
+    cdc.type = USB_TYPE_DESCRIPTOR_CDC;
+    cdc.subtype = 0x00;
+    usbus_control_slicer_put_bytes(usbus, (uint8_t*)&cdc, sizeof(cdc));
+    return sizeof(usb_desc_cdc_t);
+}
+
+static size_t _gen_full_acm_descriptor(usbus_t *usbus, void *arg)
+{
+    usbus_cdcacm_device_t *cdcacm = (usbus_cdcacm_device_t*)arg;
+    size_t total_len = 0;
+    total_len += _gen_cdc_descriptor(usbus);
+    total_len += _gen_acm_descriptor(usbus);
+    total_len += _gen_union_descriptor(usbus, cdcacm);
+    total_len += _gen_mngt_descriptor(usbus);
+    return total_len;
+}
+
+/* Submit (ACM interface in) */
+size_t usbus_cdc_acm_submit(usbus_cdcacm_device_t *cdcacm, const char *buf, size_t len)
+{
+    for (size_t i = 0; i < len; i++) {
+        tsrb_add_one(&cdcacm->tsrb, buf[i]);
+        if (buf[i] == '\n') {
+            _cdc_acm_flush(cdcacm);
+        }
+    }
+    return len;
+}
+
+/* flush event */
+void _cdc_acm_flush(usbus_cdcacm_device_t *cdcacm)
+{
+    usbus_event_post(cdcacm->usbus, &cdcacm->flush);
+}
+
+int cdc_init(usbus_t *usbus, usbus_cdcacm_device_t *handler,
+             usbus_cdcacm_cb_t cb, uint8_t *buf, size_t len)
+{
+    memset(handler, 0, sizeof(usbus_cdcacm_device_t));
+    handler->usbus = usbus;
+    tsrb_init(&handler->tsrb, buf, len);
+    handler->handler_ctrl.driver = &cdc_driver;
```

Comment:
`iface` and `flags` of the `handler_ctrl` member are never initialized. Is that right?

Reply:
Good catch! `flags` is set in the `_init` function, but `iface` is never used! I have to check if it can be removed without any future issues though